pdf_manager.py

```diff
@@ -0,0 +1,490 @@
+
+        for pagenum, page in enumerate(extract_pages(self.pdf_path)):
+            page_obj = pdf_reader.pages[pagenum]
+            (
+                page_text,
+                line_format,
+                text_from_images,
+                text_from_tables,
+                page_content,
+            ) = ([], [], [], [], [])
+            table_num, first_element, table_extraction_flag = 0, True, False
+
+            with pdfplumber.open(self.pdf_path) as pdf:
+                page_tables = pdf.pages[pagenum]
+                tables = page_tables.find_tables()
+
+                page_elements = [
+                    (element.y1, element) for element in page._objs
+                ]
+                page_elements.sort(key=lambda a: a[0], reverse=True)
+
+                unique_text_elements = set()
+                for i, (_, element) in enumerate(page_elements):
+                    if isinstance(element, LTTextBoxHorizontal):
+                        text_element = TextElement(
+                            element.get_text(),
+                            (element.x0, element.y0, element.x1, element.y1),
+                        )
+                        if text_element not in unique_text_elements:
+                            unique_text_elements.add(text_element)
+                            self.process_text_element(
+                                element,
+                                page_text,
+                                line_format,
+                                page_content,
+                                table_extraction_flag,
+                            )
+
+                    if isinstance(element, LTFigure):
+                        self.process_image_element(
+                            element,
+                            page_obj,
+                            text_from_images,
+                            page_content,
+                            page_text,
+                            line_format,
+                        )
+
+                    if isinstance(element, LTRect):
+                        if first_element and (table_num + 1) <= len(tables):
+                            (
+                                lower_side,
+                                upper_side,
+                            ) = self.process_table_element(
+                                element,
+                                pagenum,
+                                page,
+                                tables,
+                                table_num,
+                                text_from_tables,
+                                page_content,
+                                page_text,
+                                line_format,
+                            )
+                            table_extraction_flag = True
+                            first_element = False
+                        elif (
+                            element.y0 >= lower_side
+                            and element.y1 <= upper_side
```

Comment:
Where is lower_side and upper_side being defined here? This throws a "UnboundLocalError: cannot access local variable 'lower_side' where it is not associated with a value"